portal-impl/src/com/liferay/portal/tools/DBUpgrader.java

```diff
@@ -157,27 +163,18 @@ public static void upgrade() throws Exception {
 	public static void upgrade(ApplicationContext applicationContext)
 		throws Exception {
 
-		StartupHelperUtil.setUpgrading(true);
-
-		_upgradePortal();
-
-		DLFileEntryTypeLocalServiceUtil.getBasicDocumentDLFileEntryType();
-
-		_upgradeModules(applicationContext);
-
-		StoreFactory storeFactory = StoreFactory.getInstance();
+		UnsafeConsumer<ApplicationContext, Exception> unsafeConsumer =
+			appContext -> _upgrade(appContext);
 
-		if (storeFactory.getStore(PropsValues.DL_STORE_IMPL) == null) {
-			if (_log.isWarnEnabled()) {
-				_log.warn(
-					"Store \"" + PropsValues.DL_STORE_IMPL +
-						"\" is not available");
+		if (PropsValues.UPGRADE_DATABASE_MANAGED_STARTUP) {
+			if (!_upgradeWithLocking(applicationContext, unsafeConsumer)) {
```

Comment:
@LuisOrtiz89 this logic does not seem right. It is saying if the current node obtains the lock, it will do the upgrade by calling the old logic inside "_upgrade()". If the current node unable to get the lock it only runs "ModuleFrameworkUtil.registerContext(applicationContext);" ? That can not be right, it would be missing the ModuleServiceLifecycle registrations and some of the utility classes' initialization.

This cluster wide protection should only enforce the protection domain to make sure all nodes go through the upgrade stage in sequence, not to alter the logic inside the upgrade stage. Once all nodes' executions are serialized, it is the upgrade logic itself to ensure not rerunning the same upgrade again and again, which is already the case in all upgrade processes.

Just thinking about, if we are manually ensuring nodes are started one by one, the current upgrade logic without this change works perfectly fine, right? The issue is about starting nodes concurrently, and more than one nodes see tables that need upgrade which would lead to race condition. As long as this new protection is making the upgrade stage mutual exclusive, no addition changes to upgrade logic should be needed.